### PR TITLE
add recurrent neural network

### PR DESCRIPTION
**Pull request by JesusPinedaC**
This pull request introduces recurrent neural networks based on @Henrik-KM 's implementation in https://github.com/DeepTrackAI/deeplay/pull/23

## Example:

```python
rnn = dl.RecurrentNeuralNetwork(
    in_features=2,
    hidden_features=[32, 32],
    out_features=10
)
```
Every component of the network is fully flexible. For instance:

```python
rnn.blocks.layer.configure(nn.LSTM)
```

Or, in the case of bidirectional RNNs:

```python
rnn.blocks.layer.configure(bidirectional=True)

# The forward and backward directions are concatenated by the RNN layer, so the input_size of the hidden layers needs to be updated
rnn.blocks[1:].layer.configure(input_size=64)
```

@Henrik-KM, the idea is updating this implementation based on your inputs!

### CODE REVIEW

**Comment by giovannivolpe** [comment]
@BenjaminMidtvedt @Henrik-KM there are some conflicts that need to be resolved before merging this. Can you check and confirm we can proceed to merge?

File: deeplay/models/RNN.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        
+        if embedding:
+            self.embedding_dropout=torch.nn.Dropout(dropout)
+
+        if self.rnn_type == 'LSTM':
+            self.rnn_class = torch.nn.LSTM
+        elif self.rnn_type == 'GRU':
+            self.rnn_class = torch.nn.GRU
+        else:
+            self.rnn_class = torch.nn.RNN
```
Comment (BenjaminMidtvedt) [inline deeplay/models/RNN.py]: does not handle case where rnn_type is a module.

File: deeplay/models/RNN.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        """Return the dropout layers of the network. Equivalent to `.blocks.dropout`."""
+        return self.blocks.dropout
+
+
+    def __init__(
+        self, 
+        in_features: Optional[int], 
+        hidden_features: Sequence[int], 
+        out_features: Optional[int] = None,
+        rnn_type: Union[Literal['RNN', 'LSTM', 'GRU'],torch.nn.Module] = 'GRU',
```
Comment (BenjaminMidtvedt) [inline deeplay/models/RNN.py]: should probably be typed Type[torch.nn.Module]

File: deeplay/blocks/LAND.py
Comment (BenjaminMidtvedt) [inline deeplay/blocks/LAND.py:1]: rename lower case land

File: deeplay/list.py [hunk 3]
```diff
@@ -37,6 +37,18 @@ def append(self, module: DeeplayModule) -> "LayerList[T]":
...
+        if not self._has_built:
+            args = list(self._args)
+            args.insert(index, module)
+            self._args = tuple(args)
+            self.__construct__()
+
+        else:
+            super().insert(index, module)
+
+        return self
```
Comment (BenjaminMidtvedt) [inline deeplay/list.py]: revert in favor of solution in develop

File: deeplay/models/RNN.py
Comment (BenjaminMidtvedt) [inline deeplay/models/RNN.py:1]: lower case module name

File: deeplay/models/__init__.py [hunk 4]
```diff
@@ -0,0 +1 @@
+from .RNN import *
```
Comment (BenjaminMidtvedt) [inline deeplay/models/__init__.py]: imports from .py files should be explicit

File: deeplay/components/rnn.py [hunk 5]
```diff
@@ -1,147 +1,128 @@
...
-    - bidirectional (bool): If `True`, becomes a bidirectional RNN (Default: True)
-    - Embedding (nn.Embedding): Embedding layer for the input (Default: None)
+
+class Drop(nn.Module):
+    def __init__(self, p=0):
+        super(Drop, self).__init__()
+        self.p = p
+
+    def forward(self, x):
+        return nn.Dropout(p=self.p)(x[0]),x[1]
```
Comment (BenjaminMidtvedt) [inline deeplay/components/rnn.py]: Rename to make clear that this is dropout for recurrent networks

Comment (BenjaminMidtvedt) [inline deeplay/components/rnn.py]: Use nn.functional.dropout if possible.

File: deeplay/components/rnn.py
Comment (BenjaminMidtvedt) [inline deeplay/components/rnn.py:1]: What's the difference between this RNN and the one in models/rnn.py?

Comment (Henrik-KM) [reply deeplay/components/rnn.py:1]: This rnn is a generic component, i.e. it needs to be configured to work for any real application. The RNN in models/rnn uses this component to configure a directly usable network given the input parameters.

File: deeplay/components/rnn.py (hunk 5 above)
Comment (Henrik-KM) [reply deeplay/components/rnn.py]: Why is this preferred? NN.Dropout automatically disables dropout in model evaluation mode, for instance

Comment (BenjaminMidtvedt) [reply deeplay/components/rnn.py]: This implementation will not be disabled in eval mode, because it's not attached to the parent module. In that case, one should create the module in the init method

**Comment by cmanzo** [comment]
@JesusPinedaC @Henrik-KM 
I'm having some trouble with the following code:

```
import deeplay as dl
from torch.nn import CrossEntropyLoss

rnn = dl.RecurrentNeuralNetwork(
    in_features=n_features,
    hidden_features=[100],
    out_features=model_number,
)
rnn.blocks.dropout.configure(p=0.2)

model_rnn = dl.Classifier(
    model=rnn,
    loss=CrossEntropyLoss(),
    optimizer=dl.Adam(lr=0.001),
)
model_rnn.create()

trainer_rnn = dl.Trainer(max_epochs=30, accelerator="auto")
trainer_rnn.fit(model_rnn, train_loader)
```
I get `MisconfigurationException: Unknown configuration for model optimizers.` Am I doing anything wrong?

**Comment by JesusPinedaC** [comment]
> @JesusPinedaC @Henrik-KM I'm having some trouble with the following code:
> 
> ```
> import deeplay as dl
> from torch.nn import CrossEntropyLoss
> 
> rnn = dl.RecurrentNeuralNetwork(
>     in_features=n_features,
>     hidden_features=[100],
>     out_features=model_number,
> )
> rnn.blocks.dropout.configure(p=0.2)
> 
> model_rnn = dl.Classifier(
>     model=rnn,
>     loss=CrossEntropyLoss(),
>     optimizer=dl.Adam(lr=0.001),
> )
> model_rnn.create()
> 
> trainer_rnn = dl.Trainer(max_epochs=30, accelerator="auto")
> trainer_rnn.fit(model_rnn, train_loader)
> ```
> 
> I get `MisconfigurationException: Unknown configuration for model optimizers.` Am I doing anything wrong?

@cmanzo Doing

```python
model_rnn = dl.Classifier(
    model=rnn,
    loss=CrossEntropyLoss(),
    optimizer=dl.Adam(lr=0.001),
).create()
```
instead, solves the issue!

**Comment by cmanzo** [comment]
It seems that:
```
rnn = dl.RNN(
    in_features=4,
    hidden_features=[100],
    out_features=5,
    rnn_type="RNN",
    dropout=0.2,
)
```
doesn't set the dropout correctly. 

However, 
`rnn.blocks.dropout.configure(p=0.2)`
does work.

There is also some inconsistency when printing the models. For example
`print(rnn)` displays:

```
RNN(
  (blocks): LayerList(
    (0): RecurrentNeuralNetwork(
  ...
          (dropout): Layer[RecurrentDropout](p=0.2)
        )
...
    (1): MultiLayerPerceptron(
...
          (dropout): Layer[Dropout](p=0.2)
        )
```

but 

```
model_rnn = dl.Classifier(
    model=rnn,
    loss=torch.nn.CrossEntropyLoss(),
    optimizer=dl.Adam(lr=0.001),
).create()
print(model_rnn.model)
```

produces:

```
RNN(
  (blocks): LayerList(
    (0): RecurrentNeuralNetwork(
      (blocks): LayerList(
        ...
          (dropout): RecurrentDropout()
        )
 ...
      (blocks): LayerList(
        (0): LayerActivationNormalizationDropout(
 ...
          (dropout): Dropout(p=0.2, inplace=False)
        )
      )
```
However, the value is set correctly as it can be verified by:

`model_rnn.model.blocks[0].blocks[0].dropout.p`

**Comment by BenjaminMidtvedt** [comment]
@JesusPinedaC @Henrik-KM can you check the merge conflicts?

**Comment by Henrik-KM** [comment]
> @JesusPinedaC @Henrik-KM can you check the merge conflicts?

I think this PR can be closed, right @JesusPinedaC? All the changes should have been implemented in the separate RNN PR that was merged a few days ago.

**Comment by JesusPinedaC** [comment]
> > @JesusPinedaC @Henrik-KM can you check the merge conflicts?
> 
> I think this PR can be closed, right @JesusPinedaC? All the changes should have been implemented in the separate RNN PR that was merged a few days ago.

I think it was not merged into develop @Henrik-KM

**Comment by JesusPinedaC** [comment]
@Henrik-KM @BenjaminMidtvedt  Should we add unittests to test the new model?

**Comment by BenjaminMidtvedt** [comment]
@JesusPinedaC yes. Should formalize exactly the things to test but check the tests by carlo as a reference
